crates/holochain_cascade/src/lib.rs

```diff
@@ -993,19 +993,50 @@ where
         })
             .await??;
 
-        // For each response run the chain filter and check the invariants hold.
-        for response in results {
-            let result =
-                authority::get_agent_activity_query::must_get_agent_activity::filter_then_check(
-                    response,
-                )?;
-
-            // Short circuit if we have a result.
-            if matches!(result, MustGetAgentActivityResponse::Activity(_)) {
-                return Ok(result);
+        let mut merged_results = (MustGetAgentActivityResponse::EmptyRange, None);
+
+        for result in &results {
+            match (&merged_results, result) {
+                ((_, None), (_, Some(_))) => {
+                    merged_results = result.to_owned();
+                }
+                (
+                    (MustGetAgentActivityResponse::Activity(responses), chain_filter),
+                    (MustGetAgentActivityResponse::Activity(more_responses), other_chain_filter),
+                ) => {
+                    if chain_filter == other_chain_filter {
+                        let mut merged_responses = responses.clone();
+                        merged_responses.extend(more_responses.to_owned());
+                        let mut merged_activity =
+                            MustGetAgentActivityResponse::Activity(merged_responses);
+                        merged_activity.normalize();
+                        merged_results = (merged_activity, chain_filter.clone());
+                    }
+                    // If the chain filters disagree on what the filter is we
+                    // have a problem.
+                    else {
+                        merged_results = (MustGetAgentActivityResponse::IncompleteChain, None);
+                    }
+                }
+                ((MustGetAgentActivityResponse::Activity(_), _), _) => {
+                    // Noop.
+                }
+                _ => {
+                    merged_results = result.to_owned();
```

Comment:
The other two assignments to `merged_results` I can read off why they make sense. The first one is an init, the second is doing a clear merge but why we can overwrite with a single result here I'm not following.